cmd/maya-apiserver/spc-watcher/select_disk.go

```diff
@@ -0,0 +1,232 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return diskList, nil
+}
+
+// nodeDiskAlloter will try to allot nodes for pool creation as specified in
+// maxPool field of the storagepoolclaim.
+
+// For exapmle, if maxPool=5 and minPool=3, it will try to search for 5 nodes that will qualify for
+// pool provisioning. At least 3 node should qualify else pool will not be provisioned and pool creation
+// will be aborted gracefully with proper log messages.
+
+// If no minPool field is present,at least one node must qualify for pool provisioning.
+
+// nodeDiskAlloter can be made more intelligent as per the required pool constraints for alloting nodes.
+func (k *clientSet) nodeDiskAlloter(cp *v1alpha1.CasPool) ([]string, error) {
+
+	// spareAllotment holds the value for the remaining node allotments
+	// for pool provisioning.
+	var spareAllotment int
+
+	// assign maxPools to spareAllotment as right now maxPool is the number of allotments
+	// that needs to be done.
+	spareAllotment = cp.MaxPools
+	// get the labels on the basis of which disk list will be filtered
+	diskFilterLabel := diskFilterConstraint(cp.Type)
+
+	// Request kube-apiserver for the list of disk (powered by NDM)
+	// Currently, all the disks are returned,but the disk that is already a part of pool
+	// should not be returned.
+	listDisk, err := k.oecs.OpenebsV1alpha1().Disks().List(mach_apis_meta_v1.ListOptions{LabelSelector: diskFilterLabel})
+	if err != nil {
+		return nil, fmt.Errorf("error in getting the disk list:%v", err)
+	}
+	if listDisk == nil {
+		return nil, errors.New("no disk object found")
+	}
+
+	nodeDiskMap, spareAllotment := k.nodeSelector(listDisk, cp.PoolType, spareAllotment)
+	gotAllotment := cp.MaxPools - spareAllotment
+	if spareAllotment > cp.MaxPools-cp.MinPools {
+		return nil, fmt.Errorf("no node qualified for pool:only %d node could be alloted but required is %d", gotAllotment, cp.MinPools)
+	}
+	if gotAllotment < cp.MaxPools {
+		glog.Warning("partial node allotment done:spared node allotment:", spareAllotment)
+	}
+	selectedDisk := diskSelector(nodeDiskMap, cp.PoolType)
+	return selectedDisk, nil
+}
+
+// nodeSelector function will select candidate nodes that will qualify for storagepool provisioning in accordance
+// with the pool constraints.
+
+// NOTE: Not all the selected nodes may qualify.
+
+// Finally diskSelector function will vote for qualified nodes.
+
+func (k *clientSet) nodeSelector(listDisk *v1alpha1.DiskList, poolType string, spareAllotment int) (map[string]*nodeDisk, int) {
+
+	// Get the list of disk that has been used already for pool provisioning
+	spList, err := k.oecs.OpenebsV1alpha1().StoragePools().List(mach_apis_meta_v1.ListOptions{})
+	if err != nil {
+		glog.Error("unable to get the list of used disks:", err)
+	}
+	// Form a map that will hold all the used disk
+	usedDiskMap := make(map[string]int)
+	for _, sp := range spList.Items {
+		for _, usedDisk := range sp.Spec.Disks.DiskList {
+			usedDiskMap[usedDisk]++
+		}
```

Comment:
The above logic for getting used disks can be present in a separate function.

Reply:
There is a chance that the same node can be used to create multiple pools of same SPC. This has to be prevented by having a provisionedNodes filter.

Reply:
Fixed it